<!doctype html>
<html>
  <head>
    <meta charset="utf-8">
    <meta http-equiv="X-UA-Compatible" content="chrome=1">
    <title>raytiley.github.io</title>

    <link rel="stylesheet" href="/stylesheets/styles.css">
    <link rel="stylesheet" href="/stylesheets/pygment_trac.css">
    <meta name="viewport" content="width=device-width, initial-scale=1, user-scalable=no">
    <!--[if lt IE 9]>
    <script src="//html5shiv.googlecode.com/svn/trunk/html5.js"></script>
    <![endif]-->
  </head>
  <body>
    <div class="wrapper">
      <h1>  </h1>
      Posted On: February 09, 2016
      <hr />
      <p>I lost most of my day debugging a random test failure that we would sometimes see in development, but never on our CI server. I can only hope that this post will save some poor soul from the hours of cursing at @rwjblue in slack that I experienced today.</p>
<h2>TL;DR;</h2>
<p>If you need to use <code>Ember.onerror</code> guard it so it is not used in testing.</p>
<pre class="language-javascript"><code class="language-javascript"><span class="token keyword">if</span> <span class="token punctuation">(</span>Ember<span class="token punctuation">.</span>testing <span class="token operator">!==</span> <span class="token boolean">true</span><span class="token punctuation">)</span> <span class="token punctuation">{</span><br>  Ember<span class="token punctuation">.</span><span class="token function-variable function">onerror</span> <span class="token operator">=</span> <span class="token keyword">function</span><span class="token punctuation">(</span><span class="token punctuation">)</span> <span class="token punctuation">{</span> console<span class="token punctuation">.</span><span class="token function">log</span><span class="token punctuation">(</span><span class="token string">'log the fancy error'</span><span class="token punctuation">)</span><span class="token punctuation">;</span> <span class="token punctuation">}</span><br><span class="token punctuation">}</span></code></pre>
<h2>The Symptoms</h2>
<p>I've recently started working on a different project at my company. It is an Ember app that was started after the project I normally work on (If you wanna here the full story come here me talk at <a href="http://emberconf.com/speakers.html#ray-tiley">EmberConf</a>). Recently, some of the my team members and I have noticed a random test failure. Once it starts failing it continues to fail, even after refreshing the browser. Looking at the test more closely we all realize that the test is obviously broken and so we fix it and move on. We currently have three different topic branches with this random test fixed. The really weird part is this test never fails on our Continuous Integration server.</p>
<p>Finally this phantom failure bugged me enough that I decided to dig in. Poking around for a bit I was finally able to find a repeatable pattern, but I still didn't understand WTF was going on.</p>
<ul>
<li>If you ran the full test suite the test would never fail. This is why it alway passed in CI.</li>
<li>If you ran the test by itself, it would always fail.</li>
<li>Once the test failed once, you could run the entire test suite again and it would always fail.</li>
</ul>
<h2>Diagnosis</h2>
<p>I wish I could look back and find some concrete methodical way of tracking down this problem, but it just wasn't that type of bug. The error when the test failed was simple enough, <code>TypeError: this.attrs.saveItems is not a function</code>. The test is basically complaining that we were trying to call <code>saveItems</code> as a function. If you stuck a break point in there you would see that <code>saveItems</code> was undefined. We weren't passing the action when rendering the component in the integration test, and the test was failing. The thing is this test should ALWAYS fail. Why wasn't it failing on our CI servers? Why didn't it fail most times we ran it? Why did it always fail once it failed once? Do you see the need now for lots of cursing?</p>
<p>This type of problem screams of shared state. Something was happening when we ran the full test suite that masked the problem. To make it more confusing, once the error did occur, that shared state is no longer enough to prevent the error. Here are the ways I know of for a web application to maintain state.</p>
<ul>
<li>The Server - It's main job is to maintain state.</li>
<li>Cookies - Who still uses cookies?</li>
<li>LocalStorage - Long term key value store.</li>
<li>SessionStorage - Key value store that will maintain state until a browser session ends.</li>
<li>DB Stuff (IndexDB, WebSQL) - Cool stuff but I don't know too much about it.</li>
</ul>
<p>The Server is ruled out since our tests use <code>ember-cli-mirage</code> and therefore our data layer is blown away with each browser refresh. Our app also doesn't use cookies for anything. We use <code>LocalStorage</code> to persist one user setting that is accessed via an <code>Ember.Service</code>. I dug into this a bit but it was a dead end. Our app also doesn't use <code>IndexDB</code> or <code>WebSQL</code>. If you use any of these technologies extensively than you probably are well aware of the <code>Resources</code> pain in Chrome development tools. It allows you to inspect all of these technologies and manipulate their values.</p>
<p><img src="/images/resources.png" alt="Chrome Resources Pain"></p>
<p>This is where the first breakthrough was made. Turns out that QUnit uses <code>SessionStorage</code> to track failing tests. It runs those tests first so that it can &quot;Fail Fast&quot;. I always noticed that test failures seem to happen really quickly, but never before realized how QUnit was doing it. If I cleared <code>SessionStorage</code> using the chrome development tools in between test runs my test would always passed.</p>
<p>I now had a solid explanation for why the test would always fail after it failed the first time. One of the tests in our suite was changing the test environment so that the broken test could pass. If I ran the test in isolation it would fail, because the other test that was changing the environment hadn't run. QUnit, seeing this failure, would now run the test first, before any tests that could alter the environment.</p>
<p>My next breakthrough came from good old <code>Console</code>. Keeping the console open during the full test suite I noticed that there were errors being logged and the messages were the same as when the test should have failed. A useful tip is that you can click the links of the filename / line numbers in console messages to navigate to source code where that message was logged.</p>
<p><img src="/images/console.png" alt="Console Links"></p>
<p>Once I found the interesting bit of source code I set a breakpoint by clicking the line numbers in the left gutter. Setting a breakpoint is the same as putting in a <code>debugger</code> statement. It will cause the browser to halt execution (only if the development tools are open) and allow you to poke around and inspect the state of the app. Breakpoints are especially useful when the code you want to break into is in vendor code.</p>
<p><img src="/images/breakpoint.png" alt="Breakpoints"></p>
<p>Using the breakpoint it was pretty easy to see where the error was getting logged and thrown away, but how did we get here? While stopped in a breakpoint you can use the <code>Call Stack</code> on the right hand side to navigate the different frames of the call stack. This allows you to look backwards and see how the code was called. The frame before my breakpoint was in <code>backburner.js</code> which is usually a no mans land of <code>setTimeouts</code> but in this case was pretty useful. The logging code was being invoked by an <code>onError</code> method which was supplied by <code>options</code>. Hmm... a configurable <code>onerror</code>, this sounds familiar.</p>
<h2>Resolution</h2>
<p>Once I saw <code>onError</code> in backburner I slapped my forehead and searched my project for <code>Ember.onerror</code>. What I found was an <code>instance-initializer</code> that was being used to setup error tracking. If you haven't heard of <code>Ember.onerror</code> before you can read about it <a href="http://emberjs.com/api/#event_onerror">here</a>. It allows you to assign a function that gets called for any un-handled errors, including any exceptions thrown in the Run Loop. It's super useful for error tracking, which is why we were overriding it in our instance-initializer.</p>
<p>The problem, if you haven't already guessed, is it is a global handler, and doesn't get reset in between tests. Once you assign a function to <code>Ember.onerror</code> that function will always get used to handle errors until the global <code>Ember</code> object is reassigned (when you refresh the browser). Since <code>instance-initializer</code>s only get run for acceptance tests, our <code>Ember.onerror</code> only got assigned if we ran the full test suite. Because our acceptance tests ran before the the test in question, the error would get silently dropped and the test would pass. When we ran the test in isolation <code>Ember.onerror</code> was not assigned, the error would not be dropped, and the test would fail. At this point <code>Qunit</code> would re-order the tests so the failing test would always execute before the acceptance tests and keep on failing.</p>
<p>Fixing the issue was really simple. Only assign to <code>Ember.onerror</code> if we aren't testing.</p>
<pre class="language-javascript"><code class="language-javascript"><span class="token keyword">if</span> <span class="token punctuation">(</span>Ember<span class="token punctuation">.</span>testing <span class="token operator">!==</span> <span class="token boolean">true</span><span class="token punctuation">)</span> <span class="token punctuation">{</span><br>  Ember<span class="token punctuation">.</span><span class="token function-variable function">onerror</span> <span class="token operator">=</span> <span class="token keyword">function</span><span class="token punctuation">(</span><span class="token punctuation">)</span> <span class="token punctuation">{</span> console<span class="token punctuation">.</span><span class="token function">log</span><span class="token punctuation">(</span><span class="token string">'log the fancy error'</span><span class="token punctuation">)</span><span class="token punctuation">;</span> <span class="token punctuation">}</span><br><span class="token punctuation">}</span></code></pre>
<p>To see an example of how <code>Ember.onerror</code> can capture your errors and do things you don't expect check out this twiddle.</p>
<div style="position: relative; height: 0px; overflow: hidden; max-width: 100%; padding-bottom: 56.25%;"><iframe src="https://ember-twiddle.com/841e2afa42e6c1d195a0?fullScreen=true" style="position: absolute; top: 0px; left: 0px; width: 100%; height: 100%;"></iframe></div>
<hr>
<p>It is pretty much the same bug I spent my day today cursing. Inside an action we were trying to call a function that is <code>undefined</code>. Instead of logging anything to the console, it will instead give you a sweet <code>alert</code> message. Imagine instead if the <code>alert</code> was logging to a disabled error tracking service. The error would be swallowed and your test would pass, even though it should really fail due to an un-handeled error.</p>
<h2>Takeaways</h2>
<p>Evaluate your use of <code>Ember.onerror</code> and make sure it can't interfere with your tests. My recommendation is don't use <code>Ember.onerror</code> when testing, but your use case may differ from mine.</p>
<p>Hopefully some who aren't super familiar with the Web platforms rapid expansion of awesome tooling will get something out of my Chrome developer tools examples. I love working on web applications lately, the tools are just getting so awesome. Read more about the tools <a href="https://developer.chrome.com/devtools">here</a>.</p>
<p>Thanks for reading. Hopefully I'll see you at EmberConf 2016.</p>


        <div id="disqus_thread"></div>
        <script type="text/javascript">
            /* * * CONFIGURATION VARIABLES: EDIT BEFORE PASTING INTO YOUR WEBPAGE * * */
            var disqus_shortname = 'raytiley'; // required: replace example with your forum shortname

            /* * * DON'T EDIT BELOW THIS LINE * * */
            (function() {
                var dsq = document.createElement('script'); dsq.type = 'text/javascript'; dsq.async = true;
                dsq.src = '//' + disqus_shortname + '.disqus.com/embed.js';
                (document.getElementsByTagName('head')[0] || document.getElementsByTagName('body')[0]).appendChild(dsq);
            })();
        </script>
        <noscript>Please enable JavaScript to view the <a href="https://disqus.com/?ref_noscript">comments powered by Disqus.</a></noscript>
          
    </div>
    <script src="javascripts/scale.fix.js"></script>
    <script>
      (function(i,s,o,g,r,a,m){i['GoogleAnalyticsObject']=r;i[r]=i[r]||function(){
      (i[r].q=i[r].q||[]).push(arguments)},i[r].l=1*new Date();a=s.createElement(o),
      m=s.getElementsByTagName(o)[0];a.async=1;a.src=g;m.parentNode.insertBefore(a,m)
      })(window,document,'script','//www.google-analytics.com/analytics.js','ga');

      ga('create', 'UA-2186528-10', 'auto');
      ga('send', 'pageview');

    </script>
  </body>
</html>